Better support for APT in dev mode

FroMage opened the pull request:
Fixes #1502

With this, there are three supported ways of running APT:

- Use the APT processor as a provided dep
- Use the APT processor in the compiler plugin, via `annotationProcessorPaths`
- Use the APT processor in the compiler plugin, via  `annotationProcessors`

This works for Maven, but not Gradle. This needs tests, but not sure where to add them.

Most of this work required passing the right options to `javac` as part of DEV mode, as well as the proper output folder for generated sources (otherwise mvn would put them somewhere, and our DEV mode somewhere else).

FroMage committed:
IsolatedDevModeMain: make sure we don't ignore exceptions

FroMage committed:
Quarkus Java Compiler: make sure we set the source output folder

Otherwise, Maven places APT-generated sources in target/generated-sources/annotations
while quarkus:dev places them in target/classes

FroMage committed:
Support annotationProcessorPaths in DevMojo

FroMage committed:
Support annotationProcessors in DevMojo

aloubyansky commented on independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/workspace/SourceDir.java:
```diff
@@ -22,6 +26,8 @@ default boolean isOutputAvailable() {
 
     Path getOutputDir();
 
+    Path getGeneratedSourcesDir();
```
So far, this interface was meant to be mapping one src dir to an output dir.
You could represent the generated sources dir with a dedicated instance of `SourceDir` mapped to the same output dir as the main source dir. And then `DefaultArtifactSource` accepts collection sources and resources.

FroMage replied:
> You could represent the generated sources dir with a dedicated instance of SourceDir

With a boolean property `isGenerated`? This is used as output, I need to find it back when calling the compiler.

aloubyansky replied:
It seems like somehow it has to be annotation processor-specific, given that there are other generated sources that won't be relevant for the rest of this change. `isAtpGenerated()`?

FroMage replied:
Conceptually, I think this is not just another source dir.

First, because Quarkus does not have to look at it: it doesn't have to track it, it doesn't have to compile stuff that is in there.

Second, because it's actually both an output of a given source folder (APT outputs sources in there based on a regular Maven/Gradle source folder), and an input that produces bytecode to the same output folder as the specified source folder, except that input is automatically handled by javac and we don't have to look at it or invoke javac on it.

IMO, it's an intermediary by-product of the regular compilation.

If I add a dedicated `SourceFolder` then every plase we handle them, we'll have to skip them with `@isAptGenerated()` which seems like a lot of error-prone work, when really we're only using this location info to pass it from Maven (or Gradle) to javac.

I mean, if you insist, I can do this change, and you know this code/arch better, for sure, but I feel like this is not the right way to attach this info.

FroMage replied:
OK, done, and rebased.

gsmet commented:
@FroMage do you plan to continue progress on this for 3.7? 3.7.0.CR1 will be released next Wednesday and this looks like a very nice improvement.

FroMage commented:
I'm not sure I'd be able to finish this in time, the Gradle part is still entirely missing, and so are tests.

FroMage commented:
I've rebased and implemented the change. What remains is being able to test APT processors, really not sure where and how I could do that. And Gradle, and this is even harder.

Any guidance, @aloubyansky ?

aloubyansky commented:
Let's begin with some Maven tests. I suppose you want to test a project in dev mode for that.
In that case, you could create a sample project under https://github.com/quarkusio/quarkus/tree/main/integration-tests/maven/src/test/resources-filtered/projects
Look at how Quarkus and Maven plugin versions are configured to make sure you are testing the current Quarkus project version.
The test itself can be added to https://github.com/quarkusio/quarkus/blob/main/integration-tests/maven/src/test/java/io/quarkus/maven/it/DevMojoIT.java

I guess that could be a starting point.

FroMage commented:
Thanks!

FroMage commented:
Wow, there's sooo many projects in this test. Also, it's been running for an hour locally 😱 I'll run only my tests in the future 😂

aloubyansky commented:
You probably know @FroMage but just in case you can run individual tests with `mvn test -Dtest=DevMojoIT#<mytestmethod>`

FroMage commented:
Yeah, thanks.

While writing the docs, I noticed that QueryDsl is now advising yet another way to declare APT processors, via the `com.mysema.maven:apt-maven-plugin` plugin, which I was not aware of. I guess we'll also have to support this form of declaration… 🤦

https://github.com/querydsl/querydsl/tree/master/querydsl-jpa#querydsl-jpa

FroMage commented:
Actually, based on https://github.com/querydsl/apt-maven-plugin this is legacy, so let's forget it.

FroMage commented:
So, while adding a test, I wanted to see if the `annotationProcessorsPath` could now be used with version management, as discussed in https://github.com/quarkusio/quarkus/issues/37477 and realised that yes, it now works, because we're using the latest Maven Compiler Plugin, but either it's not supported in Aether, or I'm not doing it right when looking at the model in DevMojo:

```
Caused by: java.lang.IllegalArgumentException: version can neither be null, empty nor blank
    at org.apache.commons.lang3.Validate.notBlank (Validate.java:454)
    at org.apache.maven.artifact.ArtifactUtils.notBlank (ArtifactUtils.java:107)
    at org.apache.maven.artifact.ArtifactUtils.key (ArtifactUtils.java:97)
    at org.apache.maven.ReactorReader.findArtifact (ReactorReader.java:101)
    at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve (DefaultArtifactResolver.java:306)
    at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts (DefaultArtifactResolver.java:229)
    at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact (DefaultArtifactResolver.java:207)
    at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom (DefaultArtifactDescriptorReader.java:240)
    at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.readArtifactDescriptor (DefaultArtifactDescriptorReader.java:171)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.resolveCachedArtifactDescriptor (DefaultDependencyCollector.java:538)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.getArtifactDescriptorResult (DefaultDependencyCollector.java:523)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:410)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency (DefaultDependencyCollector.java:362)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process (DefaultDependencyCollector.java:349)
    at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies (DefaultDependencyCollector.java:254)
    at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies (DefaultRepositorySystem.java:309)
    at io.quarkus.maven.DevMojo.readAnnotationProcessorPaths (DevMojo.java:752)
    at io.quarkus.maven.DevMojo.setAnnotationProcessorFlags (DevMojo.java:1525)
```

FroMage commented:
I think I found a way

FroMage committed:
DevMojo: support managed dependencies for APT plugins

Related to #37477 and #37044

FroMage committed:
Test for APT processor support in DevMojo

Related to #1502

FroMage commented:
Rebased and added tests and dependency management for APT processors as supported by the latest Maven Java Compiler plugin (see #37477).

So now, all that's missing is the Gradle support. Is it required before we merge this, or can we do it in two steps?

FroMage commented:
Pushed a fix for the NPE

aloubyansky commented:
I think we can add Gradle support in a separate PR

FroMage commented:
Yeah, the Maven compiler creates it if it does not exist. I do wonder why it does not happen in my case, but I'll just create it and call it a week. https://github.com/psiroky/maven-compiler-plugin/blob/master/src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java#L757

FroMage commented:
This latest rebase and bug fix should fix the remaining Gradle test (not support for Gradle, just preventing it from crashing).

FroMage commented:
CI failure doesn't seem related, can't reproduce locally, re-running failed job.

aloubyansky commented:
> > I think we can add Gradle support in a separate PR
> 
> Fine by me. But still, who's our resident Gradle person who can help me with that? Or better yet, do it 😂

I'll give you some pointers similar to the Maven ones.
Here you can see sample projects https://github.com/quarkusio/quarkus/tree/main/integration-tests/gradle/src/main/resources (the quarkus versions gets filled in when projects are prepare for building)
And here you can see dev mode tests https://github.com/quarkusio/quarkus/tree/main/integration-tests/gradle/src/test/java/io/quarkus/gradle/devmode

Unlike `DevMojoIT`, in Gradle TS we have a class per test.
You can run them from the `gradle` module with a usual Maven command `mvn test -Dtest=<ClassName>`.

Hope that helps?

FroMage commented:
> Hope that helps?

Well, I also have to find out how to declare APT processors in Gradle, and how to extract their info, and where Gradle places generated sources. But sure, it does help, thanks :)

FroMage commented:
I suppose I should update documentation as part of this PR too.

FroMage commented:
> @FroMage do you plan to continue progress on this for 3.7? 3.7.0.CR1 will be released next Wednesday and this looks like a very nice improvement.

Well @gsmet looks like this is ready just in time. I'll merge it, and do docs and gradle support in another PR later.
